R/nimble_lm.R

```diff
@@ -78,6 +190,75 @@ embedLinesInForLoop <- function(lines, indexName, start = 1, finish) {
+
+  out <- list()  #First Element of List Random Effect Terms, 2nd Priors
+
+  #Random Intercept
+  if(term[[2]]==1 & is.factor(fr[[as.character(term[[3]])]])){
+
+    out[[1]] = substitute(EFFECTNAME[TERM[INDEX]],  #How to Keep double Brackets without Term Name in Front?
+                     list(EFFECTNAME = make.effect.name(term[[3]]), #was (effect.name)
+                          TERM = as.name(term[[3]]),
+                          INDEX = index.name))
+
+    groups <-  nlevels(as.factor((fr[[as.character(term[[3]])]])))
+
+    if(intcpt==1){
+        random <- make.random.prior(LHS2BUGSterm(make.effect.name(term[[3]]), index.name), mean = make.mean.name(term[[3]]),
+                                    precision = make.precision.name(term[[3]]))
+        prior.mu <- make.prior(make.mean.name(term[[3]]), 0)
+        }
+    else{
+        random <- make.random.prior(LHS2BUGSterm(make.effect.name(term[[3]]), index.name), mean = make.mean.name(term[[3]]),
+                                    precision = make.precision.name(term[[3]]))
+        prior.mu <- make.prior(make.mean.name(term[[3]]),"Normal")
+          }
+      random.loop <- embedLinesInForLoop(random, index.name, start = 1, finish = as.numeric(groups))
+
+    #Priors
+    prior.var <-  make.prior(make.precision.name(term[[3]]),"Gamma")
+    var.convert <- substitute(y <- 1/x, list(y = make.sigma.name(term[[3]]), x = make.precision.name(term[[3]])))
+    out[[2]] <- embedLinesInCurlyBrackets(list(random.loop, prior.var, var.convert, prior.mu))
+  }
+
+  #Random Slope
+  if (as.character(term[[2]]) %in% names(fr)){
+    out[[1]] = substitute(EFFECTNAME[TERM[INDEX]]*TERM2,  #How to Keep double Brackets without Term Name in Front?
+                          list(EFFECTNAME = make.effect.name(term[[2]]), #was (effect.name)
+                               TERM = as.name(term[[3]]),
+                               TERM2 = LHS2BUGSterm(term[[2]], index.name),
+                               INDEX = index.name))
+
+    groups <-  nlevels(as.factor((fr[[as.character(term[[3]])]])))
+
+    if(as.character(term[[2]]) %in% fixedterms){
+      random <- make.random.prior(LHS2BUGSterm(make.effect.name(term[[2]]), index.name), mean = make.mean.name(term[[2]]),
+                                  precision = make.precision.name(term[[2]]))
+      prior.mu <- make.prior(make.mean.name(term[[2]]), 0)
+    }
+    else{
+      random <- make.random.prior(LHS2BUGSterm(make.effect.name(term[[2]]), index.name), mean = make.mean.name(term[[2]]),
+                                  precision = make.precision.name(term[[2]]))
+      prior.mu <- make.prior(make.mean.name(term[[2]]),"Normal")
+    }
+
+    random.loop <- embedLinesInForLoop(random, index.name, start = 1, finish = as.numeric(groups))
+
+    #Priors
+    prior.var <-  make.prior(make.precision.name(term[[2]]),"Gamma")
+    var.convert <- substitute(y <- 1/x, list(y = make.sigma.name(term[[2]]), x = make.precision.name(term[[2]])))
```

Comment:
See comment about preferring standard deviations instead of precisions.  Also as a general note, we lean somewhat towards post-processing for deterministic computations since we can write nimbleFunctions to do them programmatically, but I know it is convenient and common practice to put them right into the BUGS code.